### PR TITLE
release in memory resource based on reference counting

### LINKED ISSUE #8230: Task schema support stops working
### Bug Description:
Last time during my work using `Theia` project I'm facing with the state when Task schema support stops working.
Please see additional info below

### Steps to Reproduce:
I don't have reliable steps to reproduce at the moment.
Looks like I face with the problem after a few cycles: 
`open tasks.json` ---> `edit/add a configuration` ---> `close tasks.json` --->... 

Page refresh resolves the problem, but then I face with the problem again...

### Additional Information
The error message is: `Unable to load schema from 'vscode://schemas/tasks': A resource provider for 'vscode://schemas/tasks' is not registered.`

tasks_schema_error1

- Operating System: Fedora-31
- Theia Version: fd239889618f16b8ac48d50a754bf363ee5230c9

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- fix #8230: release in memory resource based on reference counting. Otherwise it would be released by first dispose call.

JSON language server from VS Code extension opens documents for requested schemas which are getting released after 3 mins of being unused. On close it will call `MutableResource.dispose` leading to removing schema resources completely. With this PR we count references for each inmemory resource and does not dispose if there are still some clients, like schema updaters for example.

#### How to test

- Follow https://github.com/eclipse-theia/theia/pull/8340#pullrequestreview-464753633 for tasks. Error for settings is not a regression. You have to wait 3 minutes before opened schema documents will be removed from a cache if not touched.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
@RomanNikitenko Would you have a chance to verify it?

**Comment by akosyakov** [comment]
nvm, please don't review yet, calling `MutableResource.dispose` disposes all references, it should be done a bit more smarter

**Comment by akosyakov** [comment]
It should be good now. We return a resource representing a reference now, not an underlying data.

**Comment by RomanNikitenko** [comment]
> @RomanNikitenko Would you have a chance to verify it?

yes, will try to reproduce `tasks.json` related problems and let you know the result ASAP

File: packages/core/src/common/resource.ts [hunk 1]
```diff
@@ -221,25 +223,47 @@ export class MutableResource implements Resource {
...
     }
 }
+export class ReferenceMutableResource implements Resource {
+    constructor(protected reference: Reference<MutableResource>) { }
+
+    get uri(): URI {
+        return this.reference.object.uri;
+    }
+
+    get ononDidChangeContents(): Event<void> {
```
Comment (akosyakov) [inline packages/core/src/common/resource.ts:238]: it should be `onDidChangeContents`

**Pull request merged**